extends Node2D
class_name Sun

@export var sun_value := 25
@export var card_manager: CardManager
## é˜³å…‰å­˜åœ¨æ—¶é—´
@export var exist_time:float = 10.0
var collected := false  # æ˜¯å¦å·²è¢«ç‚¹å‡»æ”¶é›†
## ç”Ÿäº§é˜³å…‰ç§»åŠ¨çš„tween
var spawn_sun_tween:Tween


func _ready() -> void:
	card_manager = get_tree().current_scene.get_node("Camera2D/CardManager")
	# å¯åŠ¨ä¸€ä¸ª10ç§’å®šæ—¶å™¨
	await get_tree().create_timer(exist_time).timeout
	
	# å¦‚æœè¿˜æ²¡è¢«ç‚¹å‡»æ”¶é›†ï¼Œè‡ªåŠ¨é”€æ¯
	if not collected and is_instance_valid(self):
		_start_fade_out()
	
func _sun_scale(new_sun_value:int):
	var new_scale = new_sun_value/25.0
	scale = Vector2(new_scale,new_scale)


func _on_button_pressed() -> void:
	if spawn_sun_tween:
		spawn_sun_tween.kill()
	
	if collected:
		return  # é˜²æ­¢é‡å¤ç‚¹å‡»
	
	collected = true  # è®¾ç½®å·²è¢«æ”¶é›†
	var target_position = Vector2()
	SoundManager.play_other_SFX("points")
	if get_tree().current_scene is MainGameManager:
		var main_game_manager:MainGameManager = get_tree().current_scene
		target_position = main_game_manager.card_manager.marker_2d_sun_target.global_position
	
	card_manager.sun += sun_value
	var tween:Tween = get_tree().create_tween()
	# å°†èŠ‚ç‚¹ä»å½“å‰ä½ç½®ç§»åŠ¨åˆ°(100, 200)ï¼Œè€—æ—¶0.5ç§’
	tween.tween_property(self, "global_position", target_position, 0.3)
	$Button.queue_free()
	await tween.finished
	## åˆ°è¾¾ä½ç½®ï¼Œå˜å°å˜é€æ˜
	tween = create_tween()
	tween.set_parallel()
	tween.tween_property(self, "modulate:a", 0, 0.5)
	tween.tween_property(self, "scale", Vector2(0.5,0.5), 0.5)
	await tween.finished
	queue_free()
	
	

func _start_fade_out() -> void:
	var tween = get_tree().create_tween()
	tween.tween_property(self, "modulate:a", 0.0, 1.0)  # ğŸŒ«ï¸ 1ç§’æ·¡å‡º
	tween.finished.connect(func(): 
		if not collected and is_instance_valid(self):
			self.queue_free()
	)

func on_sun_tween_finished():
	if Global.auto_collect_sun:
		_on_button_pressed()
